src/main/java/org/jabref/logic/importer/fetcher/DoiFetcher.java

```diff
@@ -84,4 +100,22 @@ private void doPostCleanup(BibEntry entry) {
             return Collections.emptyList();
         }
     }
+
+    /**
+     * Returns registration agency. Null if no agency is found.
+     *
+     * @param doi the doi to be searched
+     * @throws JSONException
+     * @throws IOException
+     */
+    public String getAgency(DOI doi) throws JSONException, IOException {
+        String agency = null;
```

Comment:
We try to avoid null. Instead you can use Optionals.
https://blog.indrek.io/articles/optionals-in-java-8/

Reply:
Could be something like this?

```
public Optional<String> getAgency(DOI doi) throws JSONException, IOException {
         Optional<String> agency = Optional.empty();
         URLDownload download = new URLDownload(DOI.AGENCY_RESOLVER + "/" + doi.getDOI());
         JSONObject response = new JSONArray(download.asString()).getJSONObject(0);
         if (response != null) {
            agency = Optional.of(response.optString("RA"));
        }
        return agency;
    }
```
and the condition in performSearchById would then be 
```
if ("medra".equalsIgnoreCase(getAgency(doi.get()).get()))
```

Reply:
You need to check if the value is present first:
getAgency(...).ifPresent() && getAgency(...).get().equals(medra)

Reply:
Yes, sure, and I also needs to change `agency = Optional.of(response.optString("RA"));` to `agency = Optional.ofNullable(response.optString("RA"));` since optString could return a null.